backend/utils/pubsub.js

```diff
@@ -0,0 +1,254 @@
+
+export function createNotification (
+  type: string,
+  data: {| contractID: string, socketID: number |}
+): string {
+  return JSON.stringify({ type, ...data })
+}
+
+export function createResponse (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+/**
+ * Creates a pubsub server instance.
+ *
+ * @param {(http.Server|https.Server)} server - A Node.js HTTP/S server to attach to.
+ * @param {Object?} options
+ * {object?} clientHandlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {object?} serverHandlers - Custom handlers for server events.
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = options.clientHandlers || {}
+  server.customServerHandlers = options.serverHandlers || {}
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers[name]
+        const defaultHandler = defaultServerHandlers[name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = {
```

Comment:
I'm finding it a challenge to follow along with the code because `this` is bound in every one of the handlers, and depending on which handler I'm looking at, the type of `this` that I'm looking at changes.

Can you add a comment above each of the handler definitions what type `this` will refer to in the handlers that follow?